### ISSUE TITLE
<string>: Fast path for SSO string constructions should be optimized

### ISSUE THREAD

**Issue by BillyONeal** [microsoft/STL#472]
As reported by @miscco it's possible we might want to factor out reallocating cases in string constructors just like we did for all mutating operations in Visual Studio 2017 15.3: https://github.com/microsoft/STL/pull/467#issuecomment-579134473

Any changes in support of this item need performance tests characterizing any improvements or regressions for both small and large string sizes.

```
C:\Users\bion\Desktop>type demo.cpp
#include <string>

void example(void* p, std::string const& s) {
    ::new(p) std::string(s);
}

C:\Users\bion\Desktop>cl /EHsc /W4 /WX /O2 /c /std:c++latest /FA /nologo .\demo.cpp
demo.cpp

```

results in (after I strip file and line markers and annotate):

```
; Function compile flags: /Ogtpy
;	COMDAT ?example@@YAXPAXABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z
_TEXT	SEGMENT
_p$ = 8							; size = 4
_s$ = 12						; size = 4
?example@@YAXPAXABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z PROC ; example, COMDAT
	push	ebx
	mov	ebx, DWORD PTR _s$[esp]
	push	ebp
	push	edi
	mov	edi, DWORD PTR _p$[esp+8]
	mov	DWORD PTR [edi+16], 0       ; Fill in the default string of zeros, 0 size, 16 capacity
	mov	DWORD PTR [edi+20], 0
	cmp	DWORD PTR [ebx+20], 16			; 00000010H
	mov	ebp, DWORD PTR [ebx+16]
	jb	SHORT $LN13@example  ; Test if the source string is in small mode
	mov	ebx, DWORD PTR [ebx]
$LN13@example:
	cmp	ebp, 16	; 00000010H ; Redundant compare for small mode
	jae	SHORT $LN10@example
	movups	xmm0, XMMWORD PTR [ebx]  ; Copy over the default zeroes and 16 with
	movups	XMMWORD PTR [edi], xmm0   ; contents of the copied string that was in SSO mode
	mov	DWORD PTR [edi+16], ebp
	mov	DWORD PTR [edi+20], 15			; 0000000fH
	pop	edi
	pop	ebp
	pop	ebx
	ret	0
$LN10@example:
	push	esi                            ; Allocate memory and fill in large mode string starts here;
	mov	esi, ebp                            ; factor in to separate function?
	mov	eax, 2147483647				; 7fffffffH ; check max_size()
	or	esi, 15					; 0000000fH
	mov	ecx, edi
	cmp	esi, eax
	cmova	esi, eax
	lea	eax, DWORD PTR [esi+1]
	push	eax
	call	?allocate@?$allocator@D@std@@QAEPADI@Z	; std::allocator<char>::allocate
	lea	ecx, DWORD PTR [ebp+1]
	mov	DWORD PTR [edi], eax
	push	ecx
	push	ebx
	push	eax
	call	_memcpy
	add	esp, 12					; 0000000cH
	mov	DWORD PTR [edi+16], ebp
	mov	DWORD PTR [edi+20], esi
	pop	esi
	pop	edi
	pop	ebp
	pop	ebx
	ret	0
?example@@YAXPAXABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z ENDP ; example
_TEXT	ENDS
```

**Comment by AlexGuteniev** [comment]
> characterizing any improvements or regressions for both small and large string sizes.

large sizes appears to be irrelevant. allocator cost is greatest impact and greatest variation

**Comment by AlexGuteniev** [comment]
The source of redundant branch on length below 16 is that we check both this and other string for SSO.
It was unneccessary before constant evaluated strings. Now it may be needed

**Comment by AlexGuteniev** [comment]
@BillyONeal 
> cmp	ebp, 16	; 00000010H ; Redundant compare for small mode

Turns out not redundant, need two comparison against 16:
 * One compares *size* with 16 to see  if destination string should use SSO
 * The other compares *reserved* with 16 to see if source string uses SSO. 
Source could be smaller than 16 and still don't use SSO, if it was large, and then was reduced, or `.reserve` was called

**Comment by BillyONeal** [comment]
>The other compares reserved with 16 to see if source string uses SSO.
Source could be smaller than 16 and still don't use SSO, if it was large, and then was reduced, or .reserve was called

If the source string has SSO engaged, then we know the destination will have SSO engaged without needing another size comparison.

**Comment by BillyONeal** [comment]
That is, this item asks for this pseudocode:

if (source._Myres == 15) {
    // memcpy _String_val
    return;
}

// now we know the source is not SSO
if (source.size <= 15) {
    copy source.ptr's chars to _Bx._Buf
    size = source.size;
    capacity = 15;
    return;
}

// allocate here

**Comment by AlexGuteniev** [comment]
I got it, will try this

**Comment by AlexGuteniev** [comment]
Surprisingly, this (one branch and `memcpy`) does not help.
Ended up just omitting couple of initializers

**Comment by AlexGuteniev** [comment]
I don't see what can be done then. Maybe someone else could try again.

**Comment by BillyONeal** [comment]
> Surprisingly, this (one branch and `memcpy`) does not help.
> Ended up just omitting couple of initializers. And had to revert that as well.

You can make a different _String_val constructor that takes an int or something and doesn't do the redundant initialization to avoid the ABI issues

**Comment by AlexGuteniev** [comment]
Ok, will try that later

**Comment by AlexGuteniev** [comment]
Turns out that the generated code varies much from the benchmark.
That is maybe `memcpy` is a good thing in general, but it caused worse timing in my benchmark.
The pseudocode given by @BillyONeal shows good assembly for his `example` function, but for my case (possibly too artificial) the original code better optimizes away to nothing.

**Comment by BillyONeal** [comment]
Right, it isn't easy to analyze / show that it's better, hence just filing this item rather than making the change myself at the time :)
